Make Message Intermediate Throw Event and Message End Event Zeebe Service Tasks

smbea opened the pull request:
- Make  Message Intermediate Throw Event and Message End Event `zeebeServiceTask`
- Tweak `canCopyProperty` behaviour to ensure only message End Events and Intermediate Throw Events (with message definitions) are able to host `zeebeServiceTask` properties

Related to https://github.com/camunda/camunda-modeler/issues/2420

MaxTru commented on test/spec/extension.js:
```diff
@@ -124,6 +124,46 @@ describe('extension - can copy', function() {
+    });
+
+
+    it('should not allow on errorEndEevent', function() {
```
Small typo :-)

```suggestion
    it('should not allow on errorEndEvent', function() {
```

MaxTru commented on lib/extension.js:
```diff
@@ -6,6 +6,7 @@ var isFunction = require('min-dash').isFunction,
 
 var WILDCARD = '*';
 
+const zeebeServiceTaskProperties = ['zeebe:TaskDefinition', 'zeebe:Subscription', 'zeebe:TaskHeaders','zeebe:LoopCharacteristics'];
```
Why is it only these four properties, and not also zeebe:Input, zeebe:Output, ...?

smbea replied:
Those are all the tasks that become available when an event becomes a zeebeServiceTask. The output, for example, was already available in any event so we don’t want to remove it in case it’s not a message event

MaxTru replied:
Okay, for output I get it, but what about for instance Input? Input is available for ZeebeServiceTasks, but shall not be available to NoneEndEvents or ThrowIntermediateEvents?

smbea replied:
Ok, I think I missed Input. I review the properties and double check which ones apply for ZeebeServiceTasks but not events

MaxTru commented on lib/extension.js:
```diff
@@ -73,4 +92,19 @@ function isAllowedInParent(property, parent) {
+}
+
+function isZeebeServiceTaskSpecific(property) {
+  console.log(property);
```
```suggestion
```

MaxTru commented on test/spec/extension.js:
```diff
@@ -124,6 +124,46 @@ describe('extension - can copy', function() {
+    it('should allow on intermediateThrowEvent with messageDefinition', function() {
+
+      // given
+      var taskDefinition = moddle.create('zeebe:LoopCharacteristics'),
```
This is not a taskDefinition?
```suggestion
      var loopCharacteristics = moddle.create('zeebe:LoopCharacteristics'),
```

MaxTru commented on test/spec/extension.js:
```diff
@@ -124,6 +124,46 @@ describe('extension - can copy', function() {
+    it('should not allow on errorEndEevent', function() {
+
+      // given
+      var taskDefinition = moddle.create('zeebe:LoopCharacteristics'),
```
```suggestion
      var loopCharacteristics = moddle.create('zeebe:LoopCharacteristics'),
```

MaxTru commented on lib/extension.js:
```diff
@@ -23,13 +24,31 @@ ZeebeModdleExtension.$inject = [ 'eventBus' ];
+    }
+  }
+
+  return true;
```
Usually we don't return in `canCopyProperty` => also see how our tests check for `not.to.be.false`. Hence I think we should not change that for consistency reasons (unless there is a particular reason?)

MaxTru replied:
The sample applies to https://github.com/zeebe-io/zeebe-bpmn-moddle/pull/15/commits/61822ec0afa91ad0d842c14d5925966d94783dbc#diff-63d870fcd395c9d6f4bd53f3f4bd7a29eb165aff442eacfbe955bfdbb96f1e9eR45

smbea replied:
For this I followed the behaviour I saw on the `canCopyProperty` in the camunda-bpmn-moddle repository (see screenshots for `canHostConnector` example).

<img width="754" alt="Screenshot 2021-09-22 at 09 14 54" src="https://user-images.githubusercontent.com/25825387/134299207-bd6de1bf-9516-4a2f-9b6c-47b4031af30b.png">
<img width="531" alt="Screenshot 2021-09-21 at 20 00 06" src="https://user-images.githubusercontent.com/25825387/134299255-3095d355-c801-413e-9ad1-b01083831c21.png">

MaxTru replied:
Ok, then we are good :-)

MaxTru commented on lib/extension.js:
```diff
@@ -73,4 +92,19 @@ function isAllowedInParent(property, parent) {
+  });
+}
+
+function isZeebeServiceTaskSpecific(property) {
```
Minor, but in order to make this more understandable, I think the function actually should be called `isAllowedInZeebeServiceTask` (because a property might still be allowed in other taskTypes as well, so `specific` is not the right term here)

smbea replied:
I agree that maybe the name is not the best, but this is related to the question regarding the task array. It's the ones that are allowed for ZeebeServiceTask but that are not allowed already for all events.

barmac replied:
Proposal: Change to the name suggested by Max and document in the function why mere `is(element, "zeebe:ZeebeServiceTask")` check is not enough.

smbea commented:
Changes overview:
* fixed typos and spacing 
* changed `isZeebeServiceTaskSpecific` to `isAllowedInZeebeServiceTask` and added a comment explaining its purpose
* added Input to the `zeebeServiceTaskProperties` array 
* added test cases for Input and Output properties in NoneEvents and MessageEndEvents

MaxTru commented on test/spec/extension.js:
```diff
@@ -185,6 +225,45 @@ describe('extension - can copy', function() {
       expect(canCopyProperty).to.be.false;
     });
 
+    it('should allow on NoneEndEvents', function() {
```
IMO we should not allow Input on NoneEndEvents? Do I miss something?

smbea replied:
Oops! You are correct. I did it correctly in the assertion `expect(canCopyProperty).to.be.false;` but the description of the test was wrong. Fixed it now.

MaxTru replied:
:+1:

barmac reviewed (approved):
LGTM
